src/uct/ib/mlx5/rc/rc_mlx5_common.h

```diff
@@ -482,6 +482,41 @@ void uct_rc_mlx5_handle_unexp_rndv(uct_rc_mlx5_iface_common_t *iface,
                                    unsigned byte_len, int poll_flags);
 
 
+static UCS_F_ALWAYS_INLINE ucs_ternary_auto_value_t
+uct_rc_mlx5_dp_ordering_ooo_get(uint8_t dp_ordering_ooo,
+                                uint8_t dp_ordering_ooo_force,
+                                uct_ib_mlx5_iface_config_t *config)
+{
+    /*
+     * HCA has an mlxreg admin configuration to force enable adaptive routing
+     * (AR) or not.
+     *
+     * HCA cap/cap_2 booleans:
+     * - if dp_ordering_ooo is set, QPC/DCTC can enable AR.
+     * - if dp_ordering_ooo_force is set, QPC/DCTC can request mlxreg
+     *   configuration override, useful to force disable.
+     *
+     * QP modify behavior with returned values:
+     * - UCS_AUTO: Do not affect existing system behavior.
+     * - UCS_NO  : Force AR disabling on the QP if supported. QP modify will
+     *   return error on failure.
+     * - UCS_TRY : Set AR to enable, ignored if any failure.
+     * - UCS_YES : Force AR enabling on the QP if supported. QP modify will
+     *   return error on failure.
+     */
+
+    if ((config->ar_enable == UCS_TRY) && dp_ordering_ooo) {
+        return UCS_TRY;
+    } else if ((config->ar_enable == UCS_NO) && dp_ordering_ooo_force) {
```

Comment:
seems there is no error if user configuration is not supported. for example, user set UCS_YES to force enable, but dp_ordering_ooo=0 (which means OOO is not supported?)

Reply:
In the case you mention, if `dp_ordering_ooo_force=1`, I think it's ok, user asked to force and error if not possible, and QP modify will return error.

Generally, I was assuming that if we cannot ensure exact request, fall back to the "closest" possible. Do you want to return an error if the two flags do not exactly match the `->ar_enable` request?

Reply:
> Generally, I was assuming that if we cannot ensure exact request, fall back to the "closest" possible. Do you want to return an error if the two flags do not exactly match the `->ar_enable` request?

yes, otherwise I'm not sure there is a difference between "yes" and "try"?

Reply:
fixed, retested param handling on roce machine.